uofw/uofw
# Add Power Service

## Description
This PR adds the fully reverse-engineered Power Service (power_0x.prx) to uofw.

Except for the PSP Go (PSP N-1000), all PSP series should be supported. We currently don't have support for the PSP Go model due to a missing reverse engineering of the PSP Go's exclusive hibernation feature. I don't have any plans right now to add that support, so if anyone wants to take a look at that, feel free to go ahead :)

## Motivation and Context
Closes #39. 

## How Has This Been Tested?
The module has been compiled and tested successfully on a PSP-1000. Testing included the following: 
* Verify that the PSP still boots and runs correctly (XMB & game)
* Verify that the PSP suspends and resumes correctly
* Verify that power saving features like dimming/turning off the PSP display work
* Verify that battery data is displayed correct in the system

Specific APIs have not been tested and as such might contain bugs. However, the overall impression after testing is that our power module can be run on a PSP device without causing any major issues.

## Additional Context
Power service can be compiled using the archaic PSPSDK 0.11 release as well as the most recent PSPSDK as part of the [PSP toolchain](https://github.com/pspdev/psptoolchain). Once we've [updated the toolchain version used in our CI](https://github.com/uofw/uofw/issues/81), support for PSPSDK 0.11 can be removed from power.

Note that if you want to run the power module on your PSP using the UOFW installer, make sure to comment out its [patching of the power service](https://github.com/uofw/uofwinst/blob/master/SystemControl/syspatch.c#L239) to make sure no issue is introduced by running power through the UOFW installer.

## Final notes
I had started working on power all the way back in early 2013 and then stopped soon after that...until starting to work on it again since the beginning of this year. If it weren't for the amazing PSP community on discord, I doubt I would have ever gathered the necessary motivation and drive to finally bring power over the finish line. Many thanks to the PSP community 👍
